xdsl/dialects/mod_arith.py

```diff
@@ -0,0 +1,71 @@
+"""
+mod_arith is a dialect implementing modular arithmetic, originally
+implemented as part of the HEIR project (https://github.com/google/heir/tree/main).
+"""
+
+from abc import ABC
+from typing import Annotated, Generic, TypeVar
+
+from xdsl.dialects.arith import signlessIntegerLike
+from xdsl.dialects.builtin import AnyIntegerAttr
+from xdsl.ir import Attribute, Dialect, Operation, SSAValue
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.traits import Pure
+
+
+class ModArithOp(IRDLOperation, ABC):
+    pass
+
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class BinaryOp(ModArithOp, ABC, Generic[_T]):
+    """
+    Simple binary operation
+    """
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+    lhs = operand_def(T)
+    rhs = operand_def(T)
+    output = result_def(T)
+
+    assembly_format = "$lhs `,` $rhs attr-dict `:` type($output)"
+    traits = frozenset((Pure(),))
+
+    def __init__(
+        self,
+        lhs: SSAValue | Operation,
+        rhs: SSAValue | Operation,
+        result_type: Attribute | None,
+    ):
+        if result_type is None:
+            result_type = SSAValue.get(lhs).type
+
+        super().__init__(operands=[lhs, rhs], result_types=[result_type])
+
+
+class BinaryModuloOp(BinaryOp[_T], ABC, Generic[_T]):
```

Comment:
Why make another class here where upstream just has `ModArith_Op` > `ModArith_BinaryOp` > `ModArith_AddOp`

Reply:
Hmm, not 100% sure what you mean, but there are `ModArith` ops that don't have a modulus.

I was thinking we could eventually model `HasModulus` as a mixin instead of a subclass. I guess this becomes relevant once we get the unary and ternary operators in?

Reply:
In the file I'm looking at, all `ModArith_BinaryOp` have a modulus (maybe I'm looking at an out of date file